### DISCUSSION TITLE
U-boot FIT image

### DISCUSSION THREAD

**Discussion by R-Bas** [OE4T/meta-tegra#583] [Q&A] [answered]
Hi,

I'm trying to use a [FIT image](https://github.com/u-boot/u-boot/blob/master/doc/uImage.FIT/howto.txt) with meta-tegra on a jetson nano dev kit. The goal is to then be able to sign the image for [Verified Boot](https://github.com/u-boot/u-boot/blob/master/doc/uImage.FIT/verified-boot.txt).

I tried manually generating the FIT image and also using Poky's "kernel-fitimage" class. But, either way, I didn't find a way to successfully boot the image on the device.

Has anyone already worked with FIT images and meta-tegra ?
As this involves a lot of things outside meta-tegra is it ok to ask about this here ?
(If yes, I will post more technical details about what I've been trying until now)

Thanks

**Comment by ichergui** [comment, marked as answer]
Hey @R-Bas 
I'm working on it with @madisongh and we will make it available really soon on both platform Jetson nano and TX2.
Which Yocto release are you using ?

**Reply by R-Bas** [reply]
I'm very glad to hear that.
I'm using gatesgarth at the moment, but i can use any Yocto release as long as it is not outdated

**Reply by ichergui** [reply]
Hey @R-Bas 

FIT image support is available, please take a look [here](https://github.com/OE4T/tegra-demo-distro/wiki/Tegra-&-Uboot-Flattened-Image-Tree-(FIT-image)) and follow the steps.

For now, the complete feature is available in our stable branch `dunfell-l4t-r32.4.3 `.

Please let me know if you need help.

I will work on it to make it available in the recent Yocto release.

**Reply by R-Bas** [reply]
Hi,

I just did a minimal image build with the instructions provided and it works perfectly.

I was looking at the code and there is something i don't quite understand.
In [this commit](https://github.com/OE4T/tegra-demo-distro/commit/774629d3cbe062ac7adff450332f40ba95f0931b), do_deploy() is appended to add bootimg header to the u-boot binary when signature is enabled. Why is it only when signature is enabled that this should be done ?

**Reply by ichergui** [reply]
Hey @R-Bas 

Nice to hear that from you. 

The booting header is mandatory when signing process happen because of the `uboot dtb` concatenation with `uboot` .

We have to make it compatible with Nvidia tool .e.g. `tegrahost`  

Please take a look into the Yocto class [uboot-sign.bbclass](http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/classes/uboot-sign.bbclass?h=dunfell) to understand well.

The `uboot dtb` hold the public key that is needed for boot signature verification.

Please me know if you need more clarification.

**Reply by R-Bas** [reply]
I didn't pay attention at all to this compatibility issue when concatenating the dtb with u-boot. This is most likely why my previous attempts failed.

I've added fitImage with signature support in my own layer based on what you have done in tegrademo and everything is working fine using Poky on the Gatesgarth release.

Thank you very much !

**Reply by ichergui** [reply]
You can use [tegra-demo-distro](https://github.com/OE4T/tegra-demo-distro) - branch `gatesgarth` it will work fine as well
